### ISSUE TITLE
Server warns about opcache not running and spams log because opcache API is restricted

### ISSUE THREAD

**Issue by TDannhauer** [nextcloud/server#3602]
### Steps to reproduce
1. Install Nextcloud on PHP-FPM 7.0.15
2. Enable Opcache
3. Restrict Opcache API to some non-nextcloud files

### Expected behaviour
Nextcloud should honor that Opcache API is restricted by security reasons

### Actual behaviour
A) Log is spammed with: 'Zend OPcache API is restricted by "restrict_api" configuration directive at /<nextcloud-webroot>/lib/private/legacy/util.php#1349' and 'Zend OPcache API is restricted by "restrict_api" configuration directive at /<nextcloud-webroot>/lib/private/legacy/util.php#1317'
B) Administration page -> Serversettings -> Security warnings states that this server has no opcache running - this is wrong.

### Server configuration
**Operating system**:
Debian Stretch

**Web server:**
Apache 2.4.25

**Database:**
Postgres 9.6

**PHP version:**
7.0.15

**Nextcloud version:** (see Nextcloud admin page)
11.0.1

**Updated from an older Nextcloud/ownCloud or fresh install:**
Fresh & pristine install, default settings

**Where did you install Nextcloud from:**
Server package downloaded from nextcloud.com

**Comment by nickvergessen** [comment]
Code in question:
```
			// Zend OpCache >= 7.0.0, PHP >= 5.5.0
			if (function_exists('opcache_invalidate')) {
				$ret = opcache_invalidate($path);
			}
```

I guess the wrong admin warning is because the results can not be gathered because the methods are restricted as well.

**Comment by nickvergessen** [comment]
@MorrisJobke a simple config to mute this? Or should we really try to work around this.

@TDannhauer can you tell how you restricted it?

**Comment by TDannhauer** [comment]
Yes thats right.

 

Why tries the code to invalidate the code?

 

 

 

Von: Joas Schilling [mailto:notifications@github.com] 
Gesendet: Freitag, 24. Februar 2017 09:48
An: nextcloud/server <server@noreply.github.com>
Cc: Torben Dannhauer <torben@dannhauer.de>; Author <author@noreply.github.com>
Betreff: Re: [nextcloud/server] Server warns about opcache not running and spams log because opcache API is restricted (#3602)

 

Code in question:

                       // Zend OpCache >= 7.0.0, PHP >= 5.5.0
                       if (function_exists('opcache_invalidate')) {
                               $ret = opcache_invalidate($path);
                       }

I guess the wrong admin warning is because the results can not be gathered because the methods are restricted as well.

**Comment by TDannhauer** [comment]
Sure:

 

In the PHP-FPM pool via

 

php_admin_value[opcache.restrict_api] = /<path to opcache managing PHP scipt>

 

e.g.

php_admin_value[opcache.restrict_api] = //var/www/admin/opmonitor.php

 

In my opinion nextcloud should not do anything with the opcache. AFAIK there is no need the disable timestamp revalidation, since it is cached by the kernel und fast to query.

At least, the invalidation should only happen if opcache.validate_timestamps is set to off. If it is set to on, there is no need to invalidate.

 

Torben

 

 

Von: Joas Schilling [mailto:notifications@github.com] 
Gesendet: Freitag, 24. Februar 2017 09:51
An: nextcloud/server <server@noreply.github.com>
Cc: Torben Dannhauer <torben@dannhauer.de>; Mention <mention@noreply.github.com>
Betreff: Re: [nextcloud/server] Server warns about opcache not running and spams log because opcache API is restricted (#3602)

 

@MorrisJobke <https://github.com/MorrisJobke>  a simple config to mute this? Or should we really try to work around this.

@TDannhauer <https://github.com/TDannhauer>  can you tell how you restricted it?

**Comment by MorrisJobke** [comment]
We do this because when we write to the config.php we need to propagate this to the opcache in use. Maybe we can simply mute the warning with an `@`, @nickvergessen ?

**Comment by TDannhauer** [comment]
This would be very dirty imo. I'll provide a clean patch proposal this evening.



Von unterwegs gesendet
Sent from mobile

**Comment by nickvergessen** [comment]
@TDannhauer did you have time to look into it already?

**Comment by TDannhauer** [comment]
sorry for the delay.

Yes I looked into it. The problem is to determine whether the function opcache_invalidate() is callable.

The currently used function_exists() checks only for existence. Unfortuntely, even is_callable() ignores the restricted API. 
It seems PHP does not provide any valid procedure the verify wether the opcache api is usable without throwing any error.

Therfore the only way to find out whether opcache API is usable is be checking the configuration for "opcache.restrict_api" and compare it to the current script name. 

Since I'm not an PHP expert, I currently struggle to get my configuration value of opcache.restrict_api.
- ini_get() does not return the correct value
- get_cfg_var() does not return the correct value

The only way to figure this out was ini_get_all():

`if (strpos(ini_get_all()['opcache.restrict_api']['global_value'], $_SERVER['SCRIPT_FILENAME']) !== false)
    echo 'Allowed to use opcache API';
else
    echo 'NOT allowed to used opcache API';
`

What do you think about this approach ?
`// Zend OpCache >= 7.0.0, PHP >= 5.5.0
if (function_exists('opcache_invalidate') && strpos(ini_get_all()['opcache.restrict_api']['global_value'], $_SERVER['SCRIPT_FILENAME']) !== false)
   $ret = opcache_invalidate($path);
}`

**Comment by poVoq** [comment]
Any update on this? My shared host (OVH) also restricts this.

**Comment by arnowelzel** [comment]
The problem still exists with NextCloud 12.0.5.

Why not just putting `@` in front of it to avoid error messages as it is done with `apc_delete_file()`, `apc_clear_cache()` etc. as well? If the function can't be called due to restrictions, it makes no sense to get those errors in the logs and if you want to handle the error you should have to check the result anyway:

```
			// Zend OpCache >= 7.0.0, PHP >= 5.5.0
			if (function_exists('opcache_invalidate')) {
				$ret = @opcache_invalidate($path);
			}
```

and there is another location too, where this should be applied:

```
		// Opcache (PHP >= 5.5)
		if (function_exists('opcache_reset')) {
			@opcache_reset();
		}
```

**Comment by MorrisJobke** [comment]
Fixed by #8188
